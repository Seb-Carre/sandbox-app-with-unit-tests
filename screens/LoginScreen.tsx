import axios from "axios";
import { useRef, useState } from "react";
import { Button, Text, TextInput } from "react-native";

import { View } from "../components/Themed";

export default function LoginScreen({ navigation }) {
  const [errorMessage, setErrorMessage] = useState(
    "Your username or password is incorrect"
  );
  const allowText = true;
  const usernameRef = useRef<string>("");
  const passwordRef = useRef<string>();
  return (
    <View style={{ flex: 1, alignItems: "center", justifyContent: "center" }}>
      {allowText && <Text>Hello</Text>}
      <TextInput
        style={{ padding: 10 }}
        placeholder="Username"
        testID="Login.username"
        onChangeText={(text) => {
          usernameRef.current = text;
        }}
      ></TextInput>
      <TextInput
        style={{ padding: 10 }}
        placeholder="Password"
        testID="Login.password"
        onChangeText={(text) => {
          passwordRef.current = text;
        }}
      ></TextInput>
      <Button
        title="OK"
        testID="Login.Button"
        onPress={async () => {
          if (!passwordRef.current || !usernameRef.current) {
            setErrorMessage("Your username or password is incorrect");
          } else {
            const result = await axios.get(
              "https://jsonplaceholder.typicode.com/users/1"
            );
            if (result.data.name === "Leanne Graham") {
              navigation.navigate("NotFound");
            } else {
              navigation.navigate("NotFound");
            }
          }
        }}
      ></Button>

      {errorMessage !== "" && (
        <Text testID="Login.errorMessage">{errorMessage}</Text>
      )}
    </View>
  );
}
