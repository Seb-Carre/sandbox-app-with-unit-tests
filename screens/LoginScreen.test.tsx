import { act, fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import renderer from "react-test-renderer";

import LoginScreen from "./LoginScreen";

it("renders correctly", () => {
  // le truc de base à faire, tu devrais créer un fichier type avec tout le temps ça :)
  const tree = renderer.create(<LoginScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Renders all elements we wanted if AllowText is true", () => {
  const { getAllByText, getAllByPlaceholderText } = render(
    <LoginScreen navigation={null} />
  );

  expect(getAllByText("Hello").length).toBe(1); // si tu veux par exemple vérifier en fonction d'une valeur qu'un élément soit bien affiché ou pas, à voir si il y a pas d'autres options également

  getAllByPlaceholderText("Username");
  getAllByPlaceholderText("Password");
});

it("check if username or password are empty", async () => {
  const { getByTestId, getByText, queryByText } = render(
    <LoginScreen navigation={null} />
  );

  fireEvent.press(getByTestId("Login.Button"));

  //await waitFor(() => getByText("Your username or password is incorrect")); waitFor au cas où tu aurais besoin d'attendre une update usestate bien plus important

  //getByText("Your username or password is incorrect");

  getByText("Your username or password is incorrect");

  // await waitFor( // autre soluce également !
  //   () => {
  //     // const counter = getByTestId("Login.errorMessage");
  //     // expect(counter.props.children).toBe(
  //     //   "Your username or password is incorrect"
  //     // );
  //     getByText("Your username or password is incorrect");
  //   },
  //   { timeout: 5000 }
  // );
});

it("check if login is successful and open new url", async () => {
  const mockedNavigate = jest.fn().mockReturnValue(true);

  const mockedNavigation = { navigate: mockedNavigate };

  const { getByTestId, getByText, queryByText } = render(
    <LoginScreen navigation={mockedNavigation as any} />
  );

  fireEvent.changeText(getByTestId("Login.username"), "Eric");

  fireEvent.changeText(getByTestId("Login.password"), "Dublin");

  fireEvent.press(getByTestId("Login.Button"));

  // //Pour attendre un call API, à voir avec Axios ducoup !
  // await act(() => new Promise((resolve) => setImmediate(resolve)));

  await waitFor(
    () => {
      expect(mockedNavigate).toBeCalledWith("NotFound");
    },
    { timeout: 5000 }
  );

  // expect(mockedNavigate).toBeCalledWith("NotFound"); // vérification si cela est bien appelé avec le bon paramètre attendu
});
