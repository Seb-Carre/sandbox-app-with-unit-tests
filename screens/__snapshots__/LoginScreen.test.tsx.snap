// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text>
    Hello
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="Username"
    style={
      Object {
        "padding": 10,
      }
    }
    testID="Login.username"
  />
  <TextInput
    onChangeText={[Function]}
    placeholder="Password"
    style={
      Object {
        "padding": 10,
      }
    }
    testID="Login.password"
  />
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="Login.Button"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        OK
      </Text>
    </View>
  </View>
  <Text
    testID="Login.errorMessage"
  >
    Your username or password is incorrect
  </Text>
</View>
`;
